app/src/main/java/br/com/dcarv/arcmovies/domain/usecase/GetGenresUseCase.kt

```diff
@@ -0,0 +1,35 @@
+package br.com.dcarv.arcmovies.domain.usecase
+
+import br.com.dcarv.arcmovies.domain.abstraction.IGenreProvider
+import br.com.dcarv.arcmovies.domain.model.Genre
+import io.reactivex.Observable
+
+/**
+ * Returns a map of genres where the key is an ID and the value is an instance of
+ * {@link br.com.dcarv.arcmovies.domain.model.Genre}
+ *
+ * @author Danilo Carvalho
+ */
+class GetGenresUseCase(
+        private val provider: IGenreProvider,
+        val language: String = "pt-BR"
+) : IUseCase<Map<Long, Genre>>() {
+    private var observable : Observable<Map<Long, Genre>>? = null
```

Comment:
> What's the gain of having this observable object here acting as a cache instead of aways returning a new instance returned by the provider?
That's actually a good point. I wanted to avoid calling the API's get genre endpoint multiple times because it would very rarely change during a single session, but I noticed I used the "cache" operator on the observable, so that was unnecessary.

> Also, for the current state, do you see a way to avoid it being nullable?
Using a "lateinit var" instead and checking if it was already initialized (I remember some version of Kotlin introduced a way to check that).

Reply:
Looking at the code again, I think I stored the observable because the GetUpcomingMoviesListUseCase calls GetGenresUseCase.createObservable() and that would return a new instance of the observable. And then the API would be called again.
I'd have to run the project and test, but I think that was the reason.